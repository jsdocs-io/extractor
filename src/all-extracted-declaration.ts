import type {
  ExtractedClass,
  ExtractedClassConstructor,
  ExtractedClassMethod,
  ExtractedClassProperty,
} from "./extract-class";
import type { ExtractedEnum, ExtractedEnumMember } from "./extract-enum";
import type { ExtractedFunction } from "./extract-function";
import type {
  ExtractedInterface,
  ExtractedInterfaceCallSignature,
  ExtractedInterfaceConstructSignature,
  ExtractedInterfaceGetAccessor,
  ExtractedInterfaceIndexSignature,
  ExtractedInterfaceMethod,
  ExtractedInterfaceProperty,
  ExtractedInterfaceSetAccessor,
} from "./extract-interface";
import type { ExtractedNamespace } from "./extract-namespace";
import type { ExtractedTypeAlias } from "./extract-type-alias";
import type { ExtractedVariable } from "./extract-variable";

/**
`AllExtractedDeclaration` is the union of all possible declarations
that can be extracted, with some being found only in other declarations
(e.g., class method declarations are found only in a class declaration).
*/
export type AllExtractedDeclaration =
  | ExtractedVariable
  | ExtractedFunction
  | ExtractedClass
  | ExtractedClassConstructor
  | ExtractedClassProperty
  | ExtractedClassMethod
  | ExtractedInterface
  | ExtractedInterfaceProperty
  | ExtractedInterfaceMethod
  | ExtractedInterfaceConstructSignature
  | ExtractedInterfaceCallSignature
  | ExtractedInterfaceIndexSignature
  | ExtractedInterfaceGetAccessor
  | ExtractedInterfaceSetAccessor
  | ExtractedEnum
  | ExtractedEnumMember
  | ExtractedTypeAlias
  | ExtractedNamespace;

/**
`AllExtractedDeclarationKind` is the union of all discriminators
used to detect the kind of declaration.
*/
export type AllExtractedDeclarationKind = AllExtractedDeclaration["kind"];
