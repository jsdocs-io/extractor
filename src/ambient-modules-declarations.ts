import { ModuleDeclaration, Node, type Project } from "ts-morph";
import { isHidden } from "./is-hidden";
import { sourceFilePath } from "./source-file-path";

export type AmbientModulesDeclarationsReturn = {
  containerName: string;
  exportName: string;
  declaration: ModuleDeclaration;
}[];

export const ambientModulesDeclarations = (
  containerName: string,
  project: Project,
  pkgName?: string,
): AmbientModulesDeclarationsReturn => {
  const ambientModulesDeclarations = [];
  for (const symbol of project.getAmbientModules()) {
    for (const declaration of symbol.getDeclarations()) {
      if (isHidden(declaration) || !Node.isModuleDeclaration(declaration)) {
        continue;
      }
      if (pkgName && !sourceFilePath(declaration).startsWith(`/${pkgName}`)) {
        // Ignore ambient modules that are not from the analyzed package.
        continue;
      }
      const exportName = declaration.getName();
      // TODO:
      // Remove surrounding quotes and eventual spaces
      // const exportID = exportName
      //   .replace(/"|'/g, "")
      //   .replace(/\s/g, "_")
      //   .trim();
      ambientModulesDeclarations.push({
        containerName,
        exportName,
        declaration,
      });
    }
  }
  return ambientModulesDeclarations;
};
