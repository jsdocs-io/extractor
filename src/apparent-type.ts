import { TypeFormatFlags, ts, type Node } from "ts-morph";

export function apparentType(node: Node): string {
	// See:
	// https://github.com/dsherret/ts-morph/issues/453#issuecomment-427405736
	// https://github.com/dsherret/ts-morph/issues/453#issuecomment-667578386
	return node
		.getType()
		.getApparentType()
		.getText(
			node,
			ts.TypeFormatFlags.NoTruncation | TypeFormatFlags.UseAliasDefinedOutsideCurrentScope,
		)
		.replace(/^Number$/, "number")
		.replace(/^Boolean$/, "boolean")
		.replace(/^String$/, "string");
}
