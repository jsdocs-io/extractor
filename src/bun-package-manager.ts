import { Effect } from "effect";
import { execa } from "execa";
import { InstallPackageError, PackageManager } from "./package-manager";

/** @internal */
export const bunPackageManager = (bunPath = "bun") =>
	PackageManager.of({
		installPackage: ({ pkg, cwd }) =>
			Effect.gen(function* (_) {
				// Run `bun add <pkg> --verbose`.
				// See https://bun.sh/docs/cli/add.
				const { stdout } = yield* _(
					Effect.tryPromise({
						try: () => execa(bunPath, ["add", pkg, "--verbose"], { cwd }),
						catch: (e) => new InstallPackageError({ cause: e }),
					}),
				);

				// With verbose output on, bun prints one line per installed package
				// (e.g., "foo@1.0.0"), including all installed dependencies.
				// These lines are between the two delimiting lines found here:
				// https://github.com/oven-sh/bun/blob/972a7b7080bd3066b54dcb43e9c91c5dfa26a69c/src/install/lockfile.zig#L5369-L5370.
				const lines = stdout.split("\n");
				const beginHash = lines.findIndex((line) => line.startsWith("-- BEGIN SHA512/256"));
				const endHash = lines.findIndex((line) => line.startsWith("-- END HASH"));
				const installedPackages = lines.slice(beginHash + 1, endHash);
				return installedPackages;
			}),
	});
