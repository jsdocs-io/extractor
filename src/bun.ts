import { execa } from "execa";

/** `Bun` is a wrapper for the Bun package manager. */
export class Bun {
	#bunCmd: string;

	constructor(bunCmd = "bun") {
		this.#bunCmd = bunCmd.trim();
	}

	/**
	`add` installs the given package in the given directory and
	returns a list of all installed packages (e.g., `["foo@1.0.0"]`).

	@see {@link https://bun.sh/docs/cli/add}
	*/
	async add(pkg: string, cwd: string): Promise<string[]> {
		// Run `bun add <pkg> --verbose` in the `cwd`.
		const { stdout } = await execa(this.#bunCmd, ["add", pkg, "--verbose"], { cwd });

		// With verbose output on, bun prints one line per installed package
		// (e.g., "foo@1.0.0"), including all installed dependencies.
		// These lines are between the two delimiting lines found here:
		// https://github.com/oven-sh/bun/blob/972a7b7080bd3066b54dcb43e9c91c5dfa26a69c/src/install/lockfile.zig#L5369-L5370.
		const lines = stdout.split("\n");
		const beginHash = lines.findIndex((line) => line.startsWith("-- BEGIN SHA512/256"));
		const endHash = lines.findIndex((line) => line.startsWith("-- END HASH"));
		const installedPackages = lines.slice(beginHash + 1, endHash);
		return installedPackages;
	}
}
