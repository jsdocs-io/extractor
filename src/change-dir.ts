import { Result } from "neverthrow";
import { OsError } from "./errors";

export const changeDir = Result.fromThrowable(
  process.chdir,
  (e) => new OsError("failed to change directory", { cause: e }),
);
