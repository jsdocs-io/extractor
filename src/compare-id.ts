export type WithId = {
  id: string;
};

export const compareId = (a: WithId, b: WithId): number =>
  a.id.localeCompare(b.id);
