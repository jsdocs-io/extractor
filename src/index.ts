export {
  extractDeclarations,
  type ExtractDeclarationsOptions,
  type ExtractedDeclaration,
  type ExtractedDeclarationKind,
} from "./extract-declarations";
