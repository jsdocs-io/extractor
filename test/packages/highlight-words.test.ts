import { expect, test } from "vitest";
import { extractPackageApi } from "../../src";

test("highlight-words@1.2.2", async () => {
  expect(
    (
      await extractPackageApi({
        pkg: "highlight-words@1.2.2",
      })
    )._unsafeUnwrap(),
  ).toMatchSnapshot({
    analyzedAt: expect.any(String),
    analyzedIn: expect.any(Number),
    packages: expect.any(Array),
  });
});
