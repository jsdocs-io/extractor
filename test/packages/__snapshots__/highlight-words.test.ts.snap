// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`highlight-words@1.2.2 1`] = `
{
  "analyzedAt": Any<String>,
  "analyzedIn": Any<Number>,
  "declarations": [
    {
      "docs": [
        "/**
 * Split a text into chunks denoting which are a match and which are not based on a user search term.
 * @param text          String  The text to split.
 * @param query         String  The query to split by. This can contain multiple words.
 * @param clipBy        Number  Clip the non-matches by a certain number of words to provide context around the matches.
 * @param matchExactly  Boolean If we have multiple words in the query, we will match any of the words if exact is false. For example, searching for "brown fox" in "the brown cute fox" will yield both "brown" and "fox" as matches. While if exact is true, the same search will return no results.
 */",
      ],
      "file": "/highlight-words/dist/highlight-words.d.ts",
      "id": "function.default",
      "kind": "function",
      "line": 32,
      "name": "default",
      "signature": "default: ({
  text,
  query,
  clipBy,
  matchExactly,
}: Readonly<HighlightWords.Options>) => HighlightWords.Chunk[];",
    },
    {
      "declarations": [
        {
          "callSignatures": [],
          "constructSignatures": [],
          "docs": [],
          "file": "/highlight-words/dist/highlight-words.d.ts",
          "getAccessors": [],
          "id": "namespace.HighlightWords.interface.Chunk",
          "indexSignatures": [],
          "kind": "interface",
          "line": 2,
          "methods": [],
          "name": "Chunk",
          "properties": [
            {
              "docs": [],
              "file": "/highlight-words/dist/highlight-words.d.ts",
              "id": "namespace.HighlightWords.interface.Chunk.property.key",
              "kind": "interface-property",
              "line": 3,
              "name": "key",
              "signature": "key: string;",
            },
            {
              "docs": [],
              "file": "/highlight-words/dist/highlight-words.d.ts",
              "id": "namespace.HighlightWords.interface.Chunk.property.match",
              "kind": "interface-property",
              "line": 5,
              "name": "match",
              "signature": "match: boolean;",
            },
            {
              "docs": [],
              "file": "/highlight-words/dist/highlight-words.d.ts",
              "id": "namespace.HighlightWords.interface.Chunk.property.text",
              "kind": "interface-property",
              "line": 4,
              "name": "text",
              "signature": "text: string;",
            },
          ],
          "setAccessors": [],
          "signature": "interface Chunk {}",
        },
        {
          "callSignatures": [],
          "constructSignatures": [],
          "docs": [],
          "file": "/highlight-words/dist/highlight-words.d.ts",
          "getAccessors": [],
          "id": "namespace.HighlightWords.interface.Clip",
          "indexSignatures": [],
          "kind": "interface",
          "line": 13,
          "methods": [],
          "name": "Clip",
          "properties": [
            {
              "docs": [],
              "file": "/highlight-words/dist/highlight-words.d.ts",
              "id": "namespace.HighlightWords.interface.Clip.property.clipBy",
              "kind": "interface-property",
              "line": 17,
              "name": "clipBy",
              "signature": "clipBy?: number;",
            },
            {
              "docs": [],
              "file": "/highlight-words/dist/highlight-words.d.ts",
              "id": "namespace.HighlightWords.interface.Clip.property.curr",
              "kind": "interface-property",
              "line": 14,
              "name": "curr",
              "signature": "curr: Chunk;",
            },
            {
              "docs": [],
              "file": "/highlight-words/dist/highlight-words.d.ts",
              "id": "namespace.HighlightWords.interface.Clip.property.next",
              "kind": "interface-property",
              "line": 15,
              "name": "next",
              "signature": "next?: Chunk;",
            },
            {
              "docs": [],
              "file": "/highlight-words/dist/highlight-words.d.ts",
              "id": "namespace.HighlightWords.interface.Clip.property.prev",
              "kind": "interface-property",
              "line": 16,
              "name": "prev",
              "signature": "prev?: Chunk;",
            },
          ],
          "setAccessors": [],
          "signature": "interface Clip {}",
        },
        {
          "callSignatures": [],
          "constructSignatures": [],
          "docs": [],
          "file": "/highlight-words/dist/highlight-words.d.ts",
          "getAccessors": [],
          "id": "namespace.HighlightWords.interface.Options",
          "indexSignatures": [],
          "kind": "interface",
          "line": 7,
          "methods": [],
          "name": "Options",
          "properties": [
            {
              "docs": [],
              "file": "/highlight-words/dist/highlight-words.d.ts",
              "id": "namespace.HighlightWords.interface.Options.property.clipBy",
              "kind": "interface-property",
              "line": 10,
              "name": "clipBy",
              "signature": "clipBy?: number;",
            },
            {
              "docs": [],
              "file": "/highlight-words/dist/highlight-words.d.ts",
              "id": "namespace.HighlightWords.interface.Options.property.matchExactly",
              "kind": "interface-property",
              "line": 11,
              "name": "matchExactly",
              "signature": "matchExactly?: boolean;",
            },
            {
              "docs": [],
              "file": "/highlight-words/dist/highlight-words.d.ts",
              "id": "namespace.HighlightWords.interface.Options.property.query",
              "kind": "interface-property",
              "line": 9,
              "name": "query",
              "signature": "query: string;",
            },
            {
              "docs": [],
              "file": "/highlight-words/dist/highlight-words.d.ts",
              "id": "namespace.HighlightWords.interface.Options.property.text",
              "kind": "interface-property",
              "line": 8,
              "name": "text",
              "signature": "text: string;",
            },
          ],
          "setAccessors": [],
          "signature": "interface Options {}",
        },
        {
          "callSignatures": [],
          "constructSignatures": [],
          "docs": [],
          "file": "/highlight-words/dist/highlight-words.d.ts",
          "getAccessors": [],
          "id": "namespace.HighlightWords.interface.Query",
          "indexSignatures": [],
          "kind": "interface",
          "line": 19,
          "methods": [],
          "name": "Query",
          "properties": [
            {
              "docs": [],
              "file": "/highlight-words/dist/highlight-words.d.ts",
              "id": "namespace.HighlightWords.interface.Query.property.matchExactly",
              "kind": "interface-property",
              "line": 21,
              "name": "matchExactly",
              "signature": "matchExactly?: boolean;",
            },
            {
              "docs": [],
              "file": "/highlight-words/dist/highlight-words.d.ts",
              "id": "namespace.HighlightWords.interface.Query.property.terms",
              "kind": "interface-property",
              "line": 20,
              "name": "terms",
              "signature": "terms: string;",
            },
          ],
          "setAccessors": [],
          "signature": "interface Query {}",
        },
      ],
      "docs": [],
      "file": "/highlight-words/dist/highlight-words.d.ts",
      "id": "namespace.HighlightWords",
      "kind": "namespace",
      "line": 1,
      "name": "HighlightWords",
      "signature": "namespace HighlightWords {}",
    },
  ],
  "name": "highlight-words",
  "overview": undefined,
  "packages": Any<Array>,
  "subpath": ".",
  "types": "dist/highlight-words.d.ts",
  "version": "1.2.2",
}
`;
